File: Easy/0094_binary-tree-inorder-traversal/binary-tree-inorder-traversal.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	var dfs func(node *TreeNode)

	dfs = func(node *TreeNode) {
		if node == nil {
			return 
		}
		dfs(node.Left)
		result = append(result, node.Val)
		dfs(node.Right)
	}

	dfs(root)
  return result
}

func BuildTree(data string) *TreeNode {
	if data == "" {
		return nil
	}

	values := strings.Split(data, ",")
	if len(values) == 0 || values[0] == "null" {
		return nil
	}

	val, err := strconv.Atoi(values[0])
	if err != nil {
		return nil
	}
	root := &TreeNode{Val: val}

	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Left = &TreeNode{Val: v}
				queue = append(queue, node.Left)
			}
		}
		i++

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Right = &TreeNode{Val: v}
				queue = append(queue, node.Right)
			}
		}
		i++
	}
	return root
}


func main() {
	tests := []string{
		"1,null,2,3",
		"1,2,3,4,5,null,8,null,null,6,7,9",
		"",
		"1",
	}

	for _, tc := range tests {
		head := BuildTree(tc)
		fmt.Printf("---\nlist: [%s]\n", tc)
		fmt.Printf("result: [%v]\n", inorderTraversal(head))
	}
}

File: Easy/0094_binary-tree-inorder-traversal/binary-tree-inorder-traversal.py
#!/bin/python3

from typing import List, Optional
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorder(self, root: Optional[TreeNode], arr: List[int]) -> None:
        if not root:
            return
        self.inorder(root.left, arr)
        arr.append(root.val)
        self.inorder(root.right, arr)

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result: List[int] = []
        self.inorder(root, result)
        return result

def build_tree(data: str) -> TreeNode:
    if not data:
        return None

    values = data.split(",")
    if not values or values[0] == "null":
        return None
    
    root = TreeNode(int(values[0]))
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] != "null":
            node.left = TreeNode(int(values[i]))
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] != "null":
            node.right = TreeNode(int(values[i]))
            queue.append(node.right)
        i += 1
    return root


def main():
    tests : List[str]= [
        "1,null,2,3",
        "1,2,3,4,5,null,8,null,null,6,7,9",
        "",
        "1",
    ]

    for tc in tests:
        head = build_tree(tc)
        print(f"---\nList: [{tc}]\nresult: {Solution().inorderTraversal(head)}")

if __name__ == "__main__":
    main()
